File: Lista01/PY/ex07.py
# Definindo a classe Professor
class Professor:
    def __init__(self, nome):
        self.nome = nome
        self.escolas = []  # Lista de escolas em que o professor leciona

    # Método para associar uma escola ao professor
    def adicionar_escola(self, escola):
        if escola not in self.escolas:
            self.escolas.append(escola)
            escola.adicionar_professor(self)  # Adiciona o professor à escola
        else:
            print(f"{self.nome} já leciona na escola {escola.nome}.")

    # Método para exibir as escolas em que o professor leciona
    def exibir_escolas(self):
        print(f"{self.nome} leciona nas seguintes escolas:")
        for escola in self.escolas:
            print(f"- {escola.nome}")

# Definindo a classe Escola
class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []  # Lista de professores da escola

    # Método para adicionar um professor à escola
    def adicionar_professor(self, professor):
        if professor not in self.professores:
            self.professores.append(professor)
        else:
            print(f"{professor.nome} já é professor na escola {self.nome}.")

    # Método para exibir os professores da escola
    def exibir_professores(self):
        print(f"Professores da escola {self.nome}:")
        for professor in self.professores:
            print(f"- {professor.nome}")

# Exemplo de uso com os novos nomes
escola_geo = Escola("GEO")
escola_liceu = Escola("Liceu Paraibano")

professor_ricardo = Professor("Ricardo")
professor_simone = Professor("Simone")

# Associando professores às escolas
professor_ricardo.adicionar_escola(escola_geo)
professor_ricardo.adicionar_escola(escola_liceu)
professor_simone.adicionar_escola(escola_geo)

# Exibindo as escolas em que cada professor leciona
professor_ricardo.exibir_escolas()
professor_simone.exibir_escolas()

# Exibindo os professores de cada escola
escola_geo.exibir_professores()
escola_liceu.exibir_professores()

File: Lista01/GO/ex07.go
package main

import "fmt"

// Definindo a struct Professor
type Professor struct {
	nome string
	escolas []*Escola // Lista de escolas em que o professor leciona
}

// Adiciona uma escola à lista de escolas do professor
func (p *Professor) AdicionarEscola(e *Escola) {
	p.escolas = append(p.escolas, e)
}

// Definindo a struct Escola
type Escola struct {
	nome       string
	professores []*Professor // Lista de professores que lecionam na escola
}

// Adiciona um professor à lista de professores da escola
func (e *Escola) AdicionarProfessor(p *Professor) {
	e.professores = append(e.professores, p)
}

// Função para exibir detalhes da escola e seus professores
func (e Escola) ExibirProfessores() {
	fmt.Printf("Escola: %s\n", e.nome)
	for _, professor := range e.professores {
		fmt.Printf("Professor: %s\n", professor.nome)
	}
}

// Função para exibir detalhes do professor e suas escolas
func (p Professor) ExibirEscolas() {
	fmt.Printf("Professor: %s\n", p.nome)
	for _, escola := range p.escolas {
		fmt.Printf("Escola: %s\n", escola.nome)
	}
}

func main() {
	// Criando as escolas
	escolaA := Escola{nome: "GEO"}
	escolaB := Escola{nome: "Liceu Paraibano"}

	// Criando os professores
	professor1 := Professor{nome: "Ricardo"}
	professor2 := Professor{nome: "Simone"}

	// Associando professores a escolas
	escolaA.AdicionarProfessor(&professor1)
	escolaA.AdicionarProfessor(&professor2)
	escolaB.AdicionarProfessor(&professor1)

	// Associando escolas a professores
	professor1.AdicionarEscola(&escolaA)
	professor1.AdicionarEscola(&escolaB)
	professor2.AdicionarEscola(&escolaA)

	// Exibindo os professores de cada escola
	escolaA.ExibirProfessores()
	escolaB.ExibirProfessores()

	// Exibindo as escolas de cada professor
	professor1.ExibirEscolas()
	professor2.ExibirEscolas()
}
